### PR TITLE
Implement `useFills` hook

### LINKED ISSUE #431: Create the useFills hook [5pt]
**Feature:** [Slot and Fill](https://community.frontity.org/t/slot-and-fill/895).
**User Stories**:
  - Add additional components to the theme.
  - Substitute certain components of the theme.

[**Sprint Card**](https://github.com/orgs/frontity/projects/4#card-36476149)

### PR DESCRIPTION
**Pull request by michalczaplinski**
**What**:

Implement the `useFills` React hook which is a prerequisite for creating `<Slot/>` components for a Slot and Fill pattern. This will allow theme creators to create "slots" where the theme users can insert their "fill" components. The "fill" can be any React component.

**Why**:

To allow users to create "slots" in themes that can then be filled by "fills". More rationale[ in the Feature Discussion](https://community.frontity.org/t/slot-and-fill/895)

**How**:

Following the Implementation Proposal: https://community.frontity.org/t/slot-and-fill/895/8

**Tasks**:

- [x] Code
- [x] TypeScript
- [x] Changeset
- [x] Community discussions
- [x] TSDocs
- [x] Unit tests
- [x] Move to "frontity" package
- [x] Add namespaces to `state.fills` and `libraries.fills`
- [x] End to end tests
- [x] TypeDocs
- [x] Documentation https://github.com/frontity/gitbook-docs/pull/75 (still WIP)

Unrelated:
- ~~TypeScript tests~~
- ~~Update starter themes~~
- ~~Update other packages~~

### CODE REVIEW

File: packages/hooks/use-fills.tsx
Comment (michalczaplinski) [inline packages/hooks/use-fills.tsx]: @orballo Could you tell me if this is the right approach to typing the output of `useConnect` ?

Or should I import the types from somewhere ?

Comment (orballo) [reply packages/hooks/use-fills.tsx]: That's the right way

File: packages/hooks/package.json
Comment (luisherranz) [inline packages/hooks/package.json]: Please don't add fixed dependencies.

File: packages/hooks/use-fills.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.tsx]: This is `fill.slot`. `fill.name` doesn't exist. I have fixed it in the IP.

File: packages/hooks/use-fills.test.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.test.tsx]: This is "slot 1". We want all fills for "slot 1" beucase this component is "Slot 1".

File: packages/hooks/use-fills.test.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.test.tsx]: `props.id` is not guaranteed to exist. Let's add the fill's name as key: https://community.frontity.org/t/slot-and-fill/895/10?u=luisherranz

File: packages/hooks/use-fills.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.tsx]: This warning is not useful because the theme will declare many slots and most of them will be empty.

**Comment by michalczaplinski** [comment]
As this is nearing finish, there are still some remaining questions:

1. Do we want to implement the [data assertions](https://community.frontity.org/t/slot-and-fill/895/3?u=mmczaplinski) here as well?
2. Is the current behaviour or returning the `[]` as a "default fill" okay?
3. Are we deprecating exporting the `roots` property of the state and does that mean refactoring e.g. the GTM package?

File: packages/hooks/package.json
Comment (michalczaplinski) [inline packages/hooks/package.json:22]: This was incorrect before for some reason.

File: packages/types/src/package.ts
Comment (michalczaplinski) [inline packages/types/src/package.ts]: I'm not entirely sure if this is how I'm supposed to add the types here. @DAreRodz Do you have some experience with this?

Comment (luisherranz) [reply packages/types/src/package.ts]: You put `fills` inside `state.frontity`. This is wrong, it's not `state.frontity.fills`, it is `state.fills`.

**Comment by luisherranz** [comment]
> 1. Do we want to implement the (data assertions)[https://community.frontity.org/t/slot-and-fill/895/3?u=mmczaplinski] here as well?

No, that's not included in the Implementation Proposal: https://community.frontity.org/t/slot-and-fill/895/6

> 2. Is the current behaviour or returning the `[]` as a "default fill" okay?

Yes, if there are no fills for that slot name an empty array is fine so people can still use `fills.map`.

> 3. Are we deprecating exporting the `roots` property of the state and does that mean refactoring e.g. the GTM package?

I suggested to rename it (while keeping backward compatibility), but that's outside of the scope of the Implementation Proposal.

File: packages/types/src/package.ts
Comment (michalczaplinski) [reply packages/types/src/package.ts]: Right, my bad, that's true. 

But assuming that I put them in `state.fills` my question was more about the general idea. Is this the right place to put the types for `fills`? I think so, but would just like to confirm.

**Comment by michalczaplinski** [comment]
> No, that's not included in the Implementation Proposal: https://community.frontity.org/t/slot-and-fill/895/6

Right, I understand, but there were several updates to the implementation proposal and as the data assertions were mentioned at some point it was not 100% clear that we are not implementing them at this point 🙂

File: packages/types/src/package.ts
Comment (luisherranz) [reply packages/types/src/package.ts]: Yes, I think the placement and the types are right 👍

**Comment by michalczaplinski** [comment]
Really not sure why pushing with `--force`  closed the PR 🤷

**Comment by michalczaplinski** [comment]
The diff is showing the changes from the base branch (`use-connect`) relative to `dev` even though I'm trying to merge `use-connect` it. I'm not sure how to unfuck this.

Let's wait until the `use-connect` branch is merged into `dev`, so then we can change the base branch to `dev` and the diff should be clean.

**Comment by michalczaplinski** [comment]
FYI, The [docs](https://github.com/frontity/gitbook-docs/pull/69) are not 100% finished for this but the code can be reviewed.

File: packages/hooks/use-fills.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.tsx]: You can do this first.

File: packages/hooks/use-fills.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.tsx]: This is missing TSDocs.

Maybe explicitly define the return type as well, so we cannot break it in a refactoring.

File: packages/hooks/use-fills.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.tsx]: Instead of redeclaring the type, you can import `Package`. That way if we change the origin type, it changes here as well.

File: packages/types/src/package.ts
Comment (luisherranz) [inline packages/types/src/package.ts]: Can you add TSDocs to this file as well?

File: packages/hooks/use-fills.test.tsx
Comment (luisherranz) [inline packages/hooks/use-fills.test.tsx]: Great tests.

The only one I miss is to check if it doesn't include a fill that doesn't match the slot name.

**Review by luisherranz — CHANGES_REQUESTED** [review]
Awesome work.

I've added some details.

And as it seems that we are going to start using this package a lot, I would like to take this opportunity to mark this as side effect free.

https://webpack.js.org/guides/tree-shaking/

Could you please do so (`sideEffects: false`) and create and index file with all the exports?
Thanks!

File: .changeset/lucky-rocks-fetch.md
Comment (luisherranz) [inline .changeset/lucky-rocks-fetch.md]: Totally my fault I didn't mention this earlier, but could you please move it to the main `"frontity"` package?

**Review by luisherranz — CHANGES_REQUESTED** [review]

File: packages/hooks/use-fills.tsx
Comment (michalczaplinski) [reply packages/hooks/use-fills.tsx]: 👍   on the TSDoc.

The return type is:
Screenshot 2020-05-22 at 14 42 44

because I assign a type here: https://github.com/frontity/frontity/blob/e2056a31a608e05d58f60c7ec98ee4ec5934209e/packages/hooks/use-fills.tsx#L47-L49

I need to specify the type of that accumulator so that I get type checks inside of the `.reduce` callback.

So the type inference figures out the correct return type of `useFills` for me. Is there a problem with this approach that I'm not aware of?

File: packages/hooks/use-fills.tsx
Comment (michalczaplinski) [reply packages/hooks/use-fills.tsx]: 👍

File: packages/types/src/package.ts
Comment (michalczaplinski) [reply packages/types/src/package.ts]: I've tried to do this, but don't really see what benefit it brings to add the TSDoc here, to be honest 🤔 

TSDoc does not really have a way to "document" the interfaces. I also don't really know what I should document here? It's just a type for the `Fill`, it should be self-evident what it is.

Could you elaborate on what you had in mind?

Comment (luisherranz) [reply packages/types/src/package.ts]: Let's not go over this, please. We need to add TSDocs to all our interfaces, functions, classes, types...

Interfaces are documented in the same way than classes.

File: packages/hooks/use-fills.tsx
Comment (luisherranz) [reply packages/hooks/use-fills.tsx]: Explicit return types make impossible to break the function (type-wise) in a subsequent refactoring.

**Review by luisherranz — DISMISSED** [review]
I've done the TSDocs myself so you can get the idea of how they work. I've also added the explicit return type so you can see what I mean.

I think the only thing missing is to move this to the main `"frontity"` package and the unit test.

**Comment by luisherranz** [comment]
I'm going to take over and finish this because I also want to add namespaces to both `state.fills` and `libraries.fills` as I have explained in the [feature discussion](https://community.frontity.org/t/slot-and-fill/895/17?u=luisherranz).

**Comment by DAreRodz** [comment]
@luisherranz, I've assigned you the issue related to this PR (#431) and moved it to **in progress**.

**Comment by luisherranz** [comment]
Thanks David! I totally have to get used to the board 😅

**Comment by luisherranz** [comment]
@michalczaplinski, this is ready for code review. I cannot add you as reviewer but could you please check it out?

I saw that TS 3.9 included some breaking changes so I did a small refactor of the types. Then I saw that it also broke our `source` data types, so I removed the upgrade and fixed the version to TS 3.8 to open that in another PR.

Other than that I moved the hook to `"frontity"`, refactored to use namespaces inside `state.fills` and `libraries.fills`, added more TSDocs and some simple e2e tests.

I still want to:

- Explore how to export the markdown from TSDocs.
- Take a look at the Doc PR: https://github.com/frontity/gitbook-docs/pull/75

But the code should be fine now.

**Comment by michalczaplinski** [comment]
Just FYI:

> I still want to:
> - Take a look at the Doc PR: frontity/gitbook-docs#75

The docs are mostly there, but I think that they still miss some more details. Are you planning to you finish those?

File: packages/frontity/src/utils/slot-and-fill/use-fills.tsx
Comment (michalczaplinski) [inline packages/frontity/src/utils/slot-and-fill/use-fills.tsx]: Just a nit but wouldn't this be clearer as:
```
 (fills, namespace) => [...fills, ...Object.entries(namespace)],
```

File: packages/core/src/app/index.tsx
Comment (michalczaplinski) [inline packages/core/src/app/index.tsx:20]: So, the `Fill` components were not supposed to be placed there in the first place?

**Review by michalczaplinski — COMMENTED** [review]
I can't "approve" or request changes  because I'm the author, but I have a couple of comments.

File: lerna.json
Comment (luisherranz) [reply lerna.json:11]: We are using `changesets` instead of `lerna` to publish our packages so that is not necessary anymore.

Sorry to have done that here but something failed in lerna, I don't remember exactly what, so I removed that code to fix it and because we don't use it anymore.

File: packages/connect/src/index.d.ts
Comment (luisherranz) [reply packages/connect/src/index.d.ts]: Oh, thanks. I'll remove that.

File: packages/core/src/app/index.tsx
Comment (luisherranz) [reply packages/core/src/app/index.tsx:20]: They were. It was the first approach to integrate Slot and Fill and the way we used it in the old framework.

But then I decided to use `state.fills` and `libraries.fills` instead, which makes much more sense for Frontity.

It is explained in https://community.frontity.org/t/slot-and-fill/895

File: packages/frontity/src/utils/slot-and-fill/use-fills.tsx
Comment (luisherranz) [reply packages/frontity/src/utils/slot-and-fill/use-fills.tsx]: Oh, absolutely. I'll change it.

**Comment by luisherranz** [comment]
> Are you planning to you finish those?

Yes. But I want to do this first to export the markdown of the TSDocs and see how we can use it: https://community.frontity.org/t/code-documentation-workflow/1633/9

File: lerna.json
Comment (michalczaplinski) [reply lerna.json:11]: oh I see, thanks :)

**Comment by michalczaplinski** [comment]
By the way, just to clarify, there were just a couple of comments but it LGTM 👍

**Comment by michalczaplinski** [comment]
Oh, sorry, I've just realized that I think you forgot to include the changesets for all the other packages that you updated 🙂

**Comment by luisherranz** [comment]
> Oh, sorry, I've just realized that I think you forgot to include the changesets for all the other packages that you updated 🙂

Thanks for the heads up! I have updated/added proper changesets 👍

**Comment by luisherranz** [comment]
As the [Typedocs approach hasn't worked as we thought](https://community.frontity.org/t/adopting-a-standard-way-to-document-frontity-apis/1662/11?u=luisherranz), I'm going to continue with this PR.

**Comment by luisherranz** [comment]
I have reused [your styles](https://github.com/wp-pwa/saturn-theme/blob/36852a9d8d39302e1603b15b879e16b0f38f80a4/src/shared/components/SlotInjector/index.js#L89-L93) @darerodz 🙂

I have added:

- Support for debug mode using `state.frontity.debug = true`.
- Unique `key`s by using the namespace as well as the fill name to avoid problems in case two packages using different namespaces use the same fill name.

This should be ready to merge. @michalczaplinski feel free to review these last two improvements again if you want.

I'll finish the docs tomorrow.

**Pull request merged**
